### PR TITLE
fix(cli): fix tracking annotation diff for non-namespaced resources (#9773)

### LINKED ISSUE #9773: Wrong diff against specific revision for non-namespaced objects when using `application.resourceTrackingMethod: annotation`
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

The tracking annotation `argocd.argoproj.io/tracking-id` contains namespace for non-namespaced objects. 
Running diff against specific revision (`argocd app diff --revision`) indicates deleting the namespace 
from the tracking annotation even if no change is going to happen.

**To Reproduce**

* have latest argocd-cli installed
* create a kind cluster (or any other test cluster alternative)
```bash
kind create cluster
```

* install Argo CD
```bash
kubectl create namespace argocd
kubectl apply -n argocd -f https://raw.githubusercontent.com/argoproj/argo-cd/v2.4.2/manifests/install.yaml
```

* patch `argocd-cm` to use `application.resourceTrackingMethod: annotation`
```bash
kubectl patch configmap/argocd-cm \
  -n argocd \
  --type merge \
  -p '{"data":{"application.resourceTrackingMethod":"annotation"}}'
```

* get initial password
```bash
kubectl -n argocd get secret argocd-initial-admin-secret -o jsonpath="{.data.password}" | base64 -d; echo
```

* apply simple test application (It contains ClusterRole which is a non-namespaced kubernetes object.)
```bash
cat <<EOF | kubectl apply -f -
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: test-clusterrole
  namespace: argocd
spec:
  destination:
    namespace: default
    server: https://kubernetes.default.svc
  project: default
  source:
    path: test-clusterrole
    repoURL: https://github.com/jojand/argocd-test.git
EOF
```

* port-forward
```bash
kubectl port-forward service/argocd-server -n argocd 8080:80
```

* sync the application in UI
* app is in sync, no diff indicated

* run argocd app diff against HEAD revision (the current synced version)
```bash
argocd app diff test-clusterrole --revision=HEAD
===== rbac.authorization.k8s.io/ClusterRole /test-clusterrole ======
5c5
<     argocd.argoproj.io/tracking-id: test-clusterrole:rbac.authorization.k8s.io/ClusterRole:default/test-clusterrole
---
>     argocd.argoproj.io/tracking-id: test-clusterrole:rbac.authorization.k8s.io/ClusterRole:/test-clusterrole
```
The diff in the tracking annotations is not expected.

**Expected behavior**

```bash
argocd app diff test-clusterrole --revision=HEAD
```
No diff indicated in annotations.

**Screenshots**

**Version**
```
argocd version
argocd: v2.4.2+c6d0c8b.dirty
  BuildDate: 2022-06-22T02:07:16Z
  GitCommit: c6d0c8baaa291cd68465acd7ad6bef58b2b6f942
  GitTreeState: dirty
  GoVersion: go1.18.3
  Compiler: gc
  Platform: darwin/amd64
argocd-server: v2.4.2+c6d0c8b
  BuildDate: 2022-06-21T20:42:05Z
  GitCommit: c6d0c8baaa291cd68465acd7ad6bef58b2b6f942
  GitTreeState: clean
  GoVersion: go1.18.3
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v4.4.1 2021-11-11T23:36:27Z
  Helm Version: v3.8.1+g5cb9af4
  Kubectl Version: v0.23.1
  Jsonnet Version: v0.18.0
```

**Logs**

### PR DESCRIPTION
**Pull request by maxbrunet**
The `UnstructuredToAppInstanceValue()` function which `SetAppInstance()` passes the namespace to says the following:

https://github.com/argoproj/argo-cd/blob/b2c59014a7a39490173408087ec4a82ceecd392d/util/argo/resource_tracking.go#L116-L120

But when called from the CLI, this namespace was always an empty string:

https://github.com/argoproj/argo-cd/blob/b2c59014a7a39490173408087ec4a82ceecd392d/cmd/argocd/commands/app.go#L1122

I have tested `argocd app diff` with `--local` and `--revision`, and they do not report the annotation diff anymore. I have not been able to test `--server-side-generate` due to another bug (related to GRPC-Web in my setup: [#12032](https://togithub.com/argoproj/argo-cd/issues/12032)), but I am confident it'll behave like the others. I have not tested with `label` as tracking method (help welcome).

Fixes #9773

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: cmd/argocd/commands/app.go [hunk 1]
```diff
@@ -1026,7 +1026,7 @@ func findandPrintDiff(ctx context.Context, app *argoappv1.Application, proj *arg
 	items := make([]objKeyLiveTarget, 0)
 	if diffOptions.local != "" {
 		localObjs := groupObjsByKey(getLocalObjects(ctx, app, proj, diffOptions.local, diffOptions.localRepoRoot, argoSettings.AppLabelKey, diffOptions.cluster.Info.ServerVersion, diffOptions.cluster.Info.APIVersions, argoSettings.KustomizeOptions, argoSettings.ConfigManagementPlugins, argoSettings.TrackingMethod), liveObjs, app.Spec.Destination.Namespace)
-		items = groupObjsForDiff(resources, localObjs, items, argoSettings, app.InstanceName(argoSettings.ControllerNamespace))
+		items = groupObjsForDiff(resources, localObjs, items, argoSettings, app.InstanceName(argoSettings.ControllerNamespace), app.Spec.Destination.Namespace)
```
Comment (maxbrunet) [inline cmd/argocd/commands/app.go:1029]: The use of `app.InstanceName(argoSettings.ControllerNamespace)` as `appName` is inconsistent with the other calls to `groupObjsForDiff()` and odd to me. I cannot wrap my head around to tell if it is legacy or a mistake

https://github.com/argoproj/argo-cd/blob/4ecffdd236bffe524538b39b293de19f905e762d/pkg/apis/application/v1alpha1/types.go#L2971-L2982

If I change it to `app.Name` integration tests fail, not sure if it'll be better fixing them or leave it as:

https://github.com/argoproj/argo-cd/actions/runs/5191714335/jobs/9359975651

Shouldn't they all build the same app name one way or another?

Comment (maxbrunet) [reply cmd/argocd/commands/app.go:1029]: Fixed by #13980

**Review by crenshaw-dev — COMMENTED** [review]
I think the PR makes sense to me. Would you mind adding tests?

**Comment by maxbrunet** [comment]
> Would you mind adding tests?

I can try, could you point me where it would be with an example I could base myself on, please?

**Comment by crenshaw-dev** [comment]
So, I think a unit test would be preferable. But our current examples of unit testing the CLI are very minimal and don't involve mocking a server (as you would have to do in this case). It might be tricky. It would probably involve modifying the `New___Cmd` function to accept an optional struct full of mock interfaces for server interactions. https://github.com/argoproj/argo-cd/blob/master/cmd/argocd/commands/version_test.go

Probably the easier way to go is to write an e2e test. There are already e2e tests which invoke the `diff` command, like this one: https://github.com/argoproj/argo-cd/blob/495d09309631ef884df61bca1bffec7206a34ec8/test/e2e/app_management_test.go#L674

**Comment by maxbrunet** [comment]
Thank you @crenshaw-dev for your guidance, I have added an E2E test, please let me know what you think.

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**